README.md
# Izumi Digital Security Standard

Optimized standard for issuing and managing digital security tokens on Ethereum, based on the proposed [ERC-1594](https://github.com/ethereum/EIPs/issues/1594) core security token interface

The base ERC-20 functionality is inherited from the well-regarded ERC-20 implementation by [Open-Zeppelin](https://github.com/OpenZeppelin/openzeppelin-solidity).

## Overview

Izumi Digital Security Standard is a new smart contract standard for issuing and managing security tokens in an efficient, regulatory-compliant manner. It directly inherits both ERC-20 and ERC-1594 to ensure compatibility, while creating rooms for necessary improvements that are unique to securities. 

Izumi Standard overrides ERC-20 `transfer` to check for transfer restrictions that security laws impose with an upgradable `RegulatorService`. The service is extended to `issuance` and `redemption` (burning), which follow the same laws. On top of traditional issuer or only regulator-initiated transfers, we have left room for individual investors to interact directly with the contract by providing proofs through `transferWith`. A feature that we believe will be increasingly used in the future as transfer restrictions and private information go off-chain.

To consolidate all KYC information for different tokens, we have implemented two contratcs `StoreKYC` and `AddressIdentities` to store all relevant KYC and whitelisted addresses, respectively. The contracts will only contain non-sensitive anonymous information, such as country code and accredited status for user id, that will be used to check for transfer restrictions.

To juggle different contracts with different roles--updating KYC information, changing transfer restrictions, issuing tokens, Izumi has taken Role Based Access Control [RBAC pattern](https://docs.openzeppelin.org/docs/learn-about-access-control.html#role-based-access-control-rolessol). This way, Izumi Standard can assign different roles to different players and create an integrated eco-system that's supported by all. 

## Getting Started

Pre-requisites, in install order:
- NodeJS and NPM, node v10+. Is this is new to you, we highly recommend using a flexible NodeJS version manager. Try [NVM](https://github.com/creationix/nvm).
- Embark `npm --global install embark`
- Project dependencies `npm install` (run in root of project dir)
- Go-ethereum.
    - Mac: `brew install ethereum` ([docs here](https://github.com/ethereum/go-ethereum/wiki/Installation-Instructions-for-Mac)).
    - Arch: `pacman -S go-ethereum` ([pkg here](https://www.archlinux.org/packages/community/x86_64/go-ethereum/)).
    - Docs for more platforms, and how to build form source: [go-ethereum wiki page here](https://github.com/ethereum/go-ethereum/wiki/Building-Ethereum)

To run the contracts in a local environment, you want to run your own node, configured to run its own test-net.
Fortunately, this is completely integrated into embark, and since it's your own freshly created network, it won't take days to sync, just a few seconds to start up.

How to run your node: simply run the command `embark blockchain` (it defaults to the development environment).

You can configure balance allocation to genesis accounts here: `./blockchain.js`.
Note that the same accounts are configured to be used in the `./contracts.js`.

After starting your local blockchain, you can run `embark build` to compile the contracts, and deploy them (it defaults to development mode).




CHANGELOG.md
# Change Log

## 0.0.1 Initial work

Set up Embark workflow for smart-contract writing.


blockchain.js
module.exports = {
  // applies to all environments
  default: {
    enabled: true,
    rpcHost: "localhost", // HTTP-RPC server listening interface (default: "localhost")
    rpcPort: 8545, // HTTP-RPC server listening port (default: 8545)
    rpcCorsDomain: { // Domains from which to accept cross origin requests (browser enforced). This can also be a comma separated list
      auto: true, // When "auto" is true, Embark will automatically set the cors to the address of the webserver
      additionalCors: [] // Additional CORS domains to add to the list. If "auto" is false, only those will be added
    },
    wsRPC: true, // Enable the WS-RPC server
    wsOrigins: { // Same thing as "rpcCorsDomain", but for WS origins
      auto: true,
      additionalCors: []
    },
    wsHost: "localhost", // WS-RPC server listening interface (default: "localhost")
    wsPort: 8546 // WS-RPC server listening port (default: 8546)

    // Accounts to use as node accounts
    // The order here corresponds to the order of `web3.eth.getAccounts`, so the first one is the `defaultAccount`
    /*,accounts: [
      {
        nodeAccounts: true, // Accounts use for the node
        numAddresses: "1", // Number of addresses/accounts (defaults to 1)
        password: "config/development/devpassword" // Password file for the accounts
      },
      // Below are additional accounts that will count as `nodeAccounts` in the `deployment` section of your contract config
      // Those will not be unlocked in the node itself
      {
        privateKey: "your_private_key"
      },
      {
        privateKeyFile: "path/to/file", // Either a keystore or a list of keys, separated by , or ;
        password: "passwordForTheKeystore" // Needed to decrypt the keystore file
      },
      {
        mnemonic: "12 word mnemonic",
        addressIndex: "0", // Optional. The index to start getting the address
        numAddresses: "1", // Optional. The number of addresses to get
        hdpath: "m/44'/60'/0'/0/" // Optional. HD derivation path
      }
    ]*/
  },

  // default environment, merges with the settings in default
  // assumed to be the intended environment by `embark run` and `embark blockchain`
  development: {
    ethereumClientName: "geth", // Can be geth or parity (default:geth)
    //ethereumClientBin: "geth",  // path to the client binary. Useful if it is not in the global PATH
    networkType: "custom", // Can be: testnet, rinkeby, livenet or custom, in which case, it will use the specified networkId
    networkId: 1337, // Network id used when networkType is custom
    isDev: true, // Uses and ephemeral proof-of-authority network with a pre-funded developer account, mining enabled
    datadir: ".embark/development/datadir", // Data directory for the databases and keystore (Geth 1.8.15 and Parity 2.0.4 can use the same base folder, till now they does not conflict with each other)
    mineWhenNeeded: true, // Uses our custom script (if isDev is false) to mine only when needed
    nodiscover: true, // Disables the peer discovery mechanism (manual peer addition)
    maxpeers: 0, // Maximum number of network peers (network disabled if set to 0) (default: 25)
    proxy: true, // Proxy is used to present meaningful information about transactions
    targetGasLimit: 8000000, // Target gas limit sets the artificial target gas floor for the blocks to mine
    simulatorBlocktime: 0, // Specify blockTime in seconds for automatic mining. Default is 0 and no auto-mining.
    accounts: [
      {
        mnemonic: "cloud stage shed hard shove bench raven ostrich once lounge ridge parade",
        addressIndex: "0",
        numAddresses: "10",
        hdpath: "m/44'/60'/0'/0/",
        balance: "5 ether"
      }
    ]
  },

  // merges with the settings in default
  // used with "embark run privatenet" and/or "embark blockchain privatenet"
  privatenet: {
    networkType: "custom",
    networkId: 1337,
    isDev: false,
    datadir: ".embark/privatenet/datadir",
    // -- mineWhenNeeded --
    // This options is only valid when isDev is false. 
    // Enabling this option uses our custom script to mine only when needed.
    // Embark creates a development account for you (using `geth account new`) and funds the account. This account can be used for
    // development (and even imported in to MetaMask). To enable correct usage, a password for this account must be specified
    // in the `account > password` setting below.
    // NOTE: once `mineWhenNeeded` is enabled, you must run an `embark reset` on your dApp before running
    // `embark blockchain` or `embark run` for the first time.
    mineWhenNeeded: true,
    // -- genesisBlock --
    // This option is only valid when mineWhenNeeded is true (which is only valid if isDev is false).
    // When enabled, geth uses POW to mine transactions as it would normally, instead of using POA as it does in --dev mode.
    // On the first `embark blockchain or embark run` after this option is enabled, geth will create a new chain with a
    // genesis block, which can be configured using the `genesisBlock` configuration option below.
    genesisBlock: "config/privatenet/genesis.json", // Genesis block to initiate on first creation of a development node
    nodiscover: true,
    maxpeers: 0,
    proxy: true,
    accounts: [
      {
        nodeAccounts: true,
        password: "config/privatenet/password" // Password to unlock the account
      }
    ],
    targetGasLimit: 8000000,
    simulatorBlocktime: 0
  },

  privateparitynet: {
    ethereumClientName: "parity",
    networkType: "custom",
    networkId: 1337,
    isDev: false,
    genesisBlock: "config/privatenet/genesis-parity.json", // Genesis block to initiate on first creation of a development node
    datadir: ".embark/privatenet/datadir",
    mineWhenNeeded: false,
    nodiscover: true,
    maxpeers: 0,
    proxy: true,
    accounts: [
      {
        nodeAccounts: true,
        password: "config/privatenet/password"
      }
    ],
    targetGasLimit: 8000000,
    simulatorBlocktime: 0
  },

  // merges with the settings in default
  // used with "embark run testnet" and/or "embark blockchain testnet"
  testnet: {
    networkType: "testnet",
    syncMode: "light",
    accounts: [
      {
        nodeAccounts: true,
        password: "config/testnet/password"
      }
    ]
  },

  // merges with the settings in default
  // used with "embark run livenet" and/or "embark blockchain livenet"
  livenet: {
    networkType: "livenet",
    syncMode: "light",
    rpcCorsDomain: "http://localhost:8000",
    wsOrigins: "http://localhost:8000",
    accounts: [
      {
        nodeAccounts: true,
        password: "config/livenet/password"
      }
    ]
  }

  // you can name an environment with specific settings and then specify with
  // "embark run custom_name" or "embark blockchain custom_name"
  //custom_name: {
  //}
};


contracts.js
module.exports = {
  // default applies to all environments
  default: {
    // Blockchain node to deploy the contracts
    deployment: {
      host: "localhost", // Host of the blockchain node
      port: 8546, // Port of the blockchain node
      type: "ws" // Type of connection (ws or rpc),
      // Accounts to use instead of the default account to populate your wallet.
      // The order here corresponds to the order of `web3.eth.getAccounts`, so the first one is the `defaultAccount`
      /*,accounts: [
        {
          privateKey: "your_private_key",
          balance: "5 ether"  // You can set the balance of the account in the dev environment
                              // Balances are in Wei, but you can specify the unit with its name
        },
        {
          privateKeyFile: "path/to/file", // Either a keystore or a list of keys, separated by , or ;
          password: "passwordForTheKeystore" // Needed to decrypt the keystore file
        },
        {
          mnemonic: "12 word mnemonic",
          addressIndex: "0", // Optional. The index to start getting the address
          numAddresses: "1", // Optional. The number of addresses to get
          hdpath: "m/44'/60'/0'/0/" // Optional. HD derivation path
        },
        {
          "nodeAccounts": true // Uses the Ethereum node's accounts
        }
      ]*/
    },
    // order of connections the dapp should connect to
    dappConnection: [
      "$WEB3",  // uses pre existing web3 object if available (e.g in Mist)
      "ws://localhost:8546",
      "http://localhost:8545"
    ],

    // Automatically call `ethereum.enable` if true.
    // If false, the following code must run before sending any transaction: `await EmbarkJS.enableEthereum();`
    // Default value is true.
    // dappAutoEnable: true,

    gas: "auto",

    // Strategy for the deployment of the contracts:
    // 'explicit' will only attempt to deploy the contracts that are explicitly specified inside the contracts section.
    strategy: 'explicit',
  },

  // default environment, merges with the settings in default
  // assumed to be the intended environment by `embark run`
  development: {
    deployment: {
      // Accounts to use instead of the default account to populate your wallet.
      // The order here corresponds to the order of `web3.eth.getAccounts`, so the first one is the `defaultAccount`
      accounts: [
        {
          mnemonic: "12 word mnemonic",
          addressIndex: "0",
          numAddresses: "10",
          hdpath: "m/44'/60'/0'/0/"
        }
      ]
    },
    dappConnection: [
      "ws://localhost:8546",
      "http://localhost:8545",
      "$WEB3"  // uses pre existing web3 object if available (e.g in Mist)
    ],
    strategy: 'explicit',
    ...multipleSecurityTokenTest
  },

  // merges with the settings in default
  // used with "embark run privatenet"
  privatenet: {
  },

  // merges with the settings in default
  // used with "embark run testnet"
  testnet: {
    deployment:{
      accounts: [
        {
          mnemonic: "hero rapid silk letter vicious hire absurd high shock cart glad guard",
          addressIndex: "0",
          numAddresses: "2",
          hdpath: "m/44'/60'/0'/0/"
        }
      ],
      host: "goerli.infura.io/v3/93f7933f8eac4d6ab9ca21fc95ee03b6",
      port: false,
      protocol: 'https',
      type: "rpc"
    },
    ...multipleSecurityTokenTest
  },

  // merges with the settings in default
  // used with "embark run livenet"
  livenet: {
  },

  // you can name an environment with specific settings and then specify with
  // "embark run custom_name" or "embark blockchain custom_name"
  //custom_name: {
  //}
};
